### PR TITLE
Arnej/wire consensus 2

### PR DESCRIPTION
**Pull request by arnej27959**
I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

@havardpe please review

### CODE REVIEW

File: slobrok/src/vespa/slobrok/server/sbenv.cpp [hunk 1]
```diff
@@ -116,15 +116,23 @@ SBEnv::SBEnv(const ConfigShim &shim)
       _rpcsrvmap()
 {
     srandom(time(nullptr) ^ getpid());
-    _rpcsrvmap.proxy().registerListener(_globalVisibleHistory);
-    _rpcsrvmap.proxy().registerListener(_localRpcMonitorMap);
+    _localMonitorSubscription = MapSubscription::subscribe(_rpcsrvmap.proxy(), _globalVisibleHistory);
+    _consensusSubscription = MapSubscription::subscribe(_rpcsrvmap.proxy(), _localRpcMonitorMap);
+    _globalHistorySubscription = MapSubscription::subscribe(_localRpcMonitorMap.dispatcher(), _consensusMap);
```
Comment (havardpe) [inline slobrok/src/vespa/slobrok/server/sbenv.cpp:121]: have the _globalHistorySubscription and _localMonitorSubscription subscriptions been swapped? (got a bit confused by the names...)

File: slobrok/src/vespa/slobrok/server/sbenv.cpp [hunk 2]
```diff
@@ -116,15 +116,23 @@ SBEnv::SBEnv(const ConfigShim &shim)
...
+    auto pList = _exchanger.getPartnerList();
+    for (const auto & partner : pList) {
+        _exchanger.removePartner(partner);
+    }
+}
 
 SBEnv::~SBEnv()
 {
     getTransport()->WaitFinished();
+    cleanup();
```
Comment (havardpe) [inline slobrok/src/vespa/slobrok/server/sbenv.cpp:135]: is the double cleanup needed? (I would expect that WaitFinished blocks forever in the cases where the other cleanup was not called...)
consider doing cleanup only here, but removing the WaitFinished call.

**Review by havardpe — APPROVED** [review]

File: slobrok/src/vespa/slobrok/server/sbenv.cpp (hunk 1 above)
Comment (arnej27959) [reply slobrok/src/vespa/slobrok/server/sbenv.cpp:121]: > have the _globalHistorySubscription and _localMonitorSubscription subscriptions been swapped?

not sure how I managed to swap these, should be fixed now.

File: slobrok/src/vespa/slobrok/server/sbenv.cpp (hunk 2 above)
Comment (arnej27959) [reply slobrok/src/vespa/slobrok/server/sbenv.cpp:135]: we expect the transport thread(s) to be finished already when we get here, yes - can we assert on that?
cleanup() is removed again, it was added to ensure that MapListeners were unregistered before destructors.

**Pull request merged**